eli5/keras.py

```diff
@@ -0,0 +1,465 @@
+        where batch size is 1 for a single image.
+
+    target_names : list, optional
+        *Not Implemented*. 
+
+        Names for classes in the final output layer.
+
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *Currently only the first prediction from the list is explained*. 
+        The list must be length one.
+
+        If None, the model is fed the input image and its top prediction 
+        is taken as the target automatically.
+
+    layer : int or str or keras.layers.Layer, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of a Keras layer.
+        
+        If None, a suitable layer is attempted to be retrieved. 
+        See :func:`eli5.keras.search_layer_backwards` for details.
+
+    Returns
+    -------
+    expl : Explanation
+        A :class:`eli5.base.Explanation` object 
+        with the ``image`` and ``heatmap`` attributes set.
+    """
+    # TODO: implement target_names
+    # FIXME: Could doc be a Tensorflow object, not just a numpy array?
+    validate_doc(estimator, doc)
+    activation_layer = get_activation_layer(estimator, layer)
+    
+    heatmap, predicted_idx, score = grad_cam(estimator, doc, targets, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations'
+    # (the output is not yet a heat map)
+
+    print('Predicted class: %d' % predicted_idx)
+    print('With probability: %f' % score)
+
+    # TODO: consider passing multiple images in doc to perform grad-cam on multiple images
+    doc = doc[0] # rank 4 batch -> rank 3 single image
+    image = keras.preprocessing.image.array_to_img(doc) # -> PIL image
+
+    return Explanation(
+        estimator.name, # might want to replace this with something else, eg: estimator.summary()
+        description=DESCRIPTION_KERAS,
+        error='',
+        method='Grad-CAM',
+        is_regression=False, # TODO: classification vs regression model
+        highlight_spaces=None, # might be relevant later when explaining text models
+        image=image, # PIL image
+        heatmap=heatmap, # 2D [0, 1] numpy array
+    )
+
+
+def validate_doc(estimator, doc):
+    # type: (Model, np.ndarray) -> None
+    """
+    Check that input ``doc`` is suitable for ``estimator``.
+
+    We assume that ``doc`` is an image.
+
+    Parameters
+    ----------
+    estimator : keras.models.Model
+        The model that will be fed the input.
+        See :func:`explain_prediction_keras`.
+
+    doc : numpy.ndarray
+        The input to be checked.
+
+
+    :raises ValueError: if ``doc`` shape does not match.
+    """
+    input_sh = estimator.input_shape
+    doc_sh = doc.shape
+    if len(input_sh) == 4:
+        # rank 4 with (batch, ...) shape
+        # check that we have only one image (batch size 1)
+        single_batch = (1, input_sh[1], input_sh[2], input_sh[3])
```

Comment:
Might be a bit cleaner with ``(1,) + input_sh[1:]``

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/80ff9ec716f86829d22918440d0a97913813da37